hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/impl/BlockRecordManagerImpl.java

```diff
@@ -0,0 +1,397 @@
+    }
+
+    // ========================================================================================================
+    // Running Hash Getter Methods
+
+    /**
+     * Get the runningHash of all RecordStreamObject. This will block if the running hash has not yet
+     * been computed for the most recent user transaction.
+     *
+     * @return the runningHash of all RecordStreamObject, or null if there are no running hashes yet
+     */
+    @NonNull
+    @Override
+    public Bytes getRunningHash() {
+        return streamFileProducer.getRunningHash();
+    }
+
+    /**
+     * Get the previous, previous, previous runningHash of all RecordStreamObject. This will block if
+     * the running hash has not yet been computed for the most recent user transaction.
+     *
+     * @return the previous, previous, previous runningHash of all RecordStreamObject, or null if there is not one yet
+     */
+    @Nullable
+    @Override
+    public Bytes getNMinus3RunningHash() {
+        return streamFileProducer.getNMinus3RunningHash();
+    }
+
+    // ========================================================================================================
+    // BlockRecordInfo Implementation
+
+    /**
+     * Get the last block number, this is the last completed immutable block.
+     *
+     * @return the current block number, 0 of there is no blocks yet
+     */
+    @Override
+    public long lastBlockNo() {
+        return lastBlockInfo.lastBlockNumber();
+    }
+
+    /**
+     * Get the consensus time of the first transaction of the last block, this is the last completed immutable block.
+     *
+     * @return the consensus time of the first transaction of the last block, null if there was no previous block
+     */
+    @Nullable
+    @Override
+    public Instant firstConsTimeOfLastBlock() {
+        final var firstConsTimeOfLastBlock = lastBlockInfo.firstConsTimeOfLastBlock();
+        return firstConsTimeOfLastBlock != null
+                ? Instant.ofEpochSecond(firstConsTimeOfLastBlock.seconds(), firstConsTimeOfLastBlock.nanos())
+                : null;
+    }
+
+    /**
+     * Gets the hash of the last block
+     *
+     * @return the last block hash, null if no blocks have been created
+     */
+    @Nullable
+    @Override
+    public Bytes lastBlockHash() {
+        return getLastBlockHash(lastBlockInfo);
+    }
+
+    /**
+     * Returns the hash of the given block number, or {@code null} if unavailable.
+     *
+     * @param blockNo the block number of interest, must be within range of (current_block - 1) -> (current_block - 254)
+     * @return its hash, if available otherwise null
+     */
+    @Nullable
+    @Override
+    public Bytes blockHashByBlockNumber(final long blockNo) {
+        final Bytes blockHashes = lastBlockInfo.blockHashes();
+        final long blocksAvailable = blockHashes.length() / HASH_SIZE;
+        if (blockNo < 0) {
+            return null;
```

Comment:
Actually, this is intentional. This API is indirectly called by smart contracts. I think throwing an exception here could be more dangerous because the interface between the smart contract systems and this code would need to handle that exception. I will add this comment:

```
        // Smart contracts (and other services) call this API. Should a smart contract call this, we don't really
        // want to throw an exception. So we will just return null, which is also valid. Basically, if the block
        // doesn't exist, you get null.
        if (blockNo < 0) {
            return null;
        }
```